### PR TITLE
TLS: support for extracting certificate subject alt names from client certs

### LINKED ISSUE #1628: TLS: add interface to query subject alternative names for client auth certificated
To augment existing DN subject info. Subject alt names can contain more, and more varied info. 
Required by certificate auth feature of scylla.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #1628
    
Subject alt name info can contain more extensive and detailed info on a connecting client. Add interface to allow querying this from a connected socket.

Initially we don't include this in accept-sequence auth verification, though we maybe should include it as an option.

### CODE REVIEW

File: src/net/tls.cc [hunk 1]
```diff
@@ -1533,25 +1533,110 @@ class session : public enable_lw_shared_from_this<session> {
...
+        if (_shutdown) {
+            return make_exception_future<result_t>(std::system_error(ENOTCONN, std::system_category()));
+        }
+        if (!_connected) {
+            return handshake().then([this, types = std::move(types)]() mutable {
+               return get_alt_name_information(std::move(types));
+            });
+        }
+
+        auto gris = extract_dn_information();
```
Comment (avikivity) [inline src/net/tls.cc:1551]: gris?

File: src/net/tls.cc [hunk 2]
```diff
@@ -1533,25 +1533,110 @@ class session : public enable_lw_shared_from_this<session> {
...
+            }
+            if (err != GNUTLS_E_SHORT_MEMORY_BUFFER) {
+                gtls_chk(err); // will throw
+            }
+            sstring buf;
+            buf.resize(size);
+
+            err = gnutls_x509_crt_get_subject_alt_name(peer.get(), i, buf.data(), &size, nullptr);
+            if (err < 0) {
+                gtls_chk(err); // will throw
```
Comment (avikivity) [inline src/net/tls.cc:1575]: Better to wrap the whole function body in futurize_invoke so the exception is made part of the future, leading to fewer surprises.

File: src/net/tls.cc (hunk 1 above)
Comment (elcallio) [reply src/net/tls.cc:1551]: Oops. Debug line that made it into commit...

**Comment by avikivity** [comment]
Looks good, but please elaborate on "Initially we don't include this in accept-sequence auth verification, though we maybe should include it as an option.". What does that mean? That we should accept some predicate to apply to alt names?

**Comment by elcallio** [comment]
It means that we should maybe add it as an (optional), retrievable attribute for the `void set_dn_verification_callback(dn_callback);`. That is a callback that can be registered for an accepting socket, and allows user code to optionally reject connections early based on certificate content (DN info). Since SAN (alt names) are an extension of DN info, it seems logical that we should maybe allw it also to be at least queried. 

I was thinking something like: 

```
    using dn_callback = noncopyable_function<void(session_type type, sstring subject, sstring issuer)>;

    void set_dn_verification_callback(dn_callback);

...

    using dn_callback_ex = noncopyable_function<void(session_type type, sstring subject, sstring issuer, const std::vector<subject_alt_name>&)>;

    void set_dn_verification_callback_ex(dn_callback_ex, std::unordered_set<subject_alt_name_type> types = {});
```

to maintain compatibility.

File: src/net/tls.cc (hunk 2 above)
Comment (elcallio) [reply src/net/tls.cc]: Fixed.

File: src/net/tls.cc (hunk 1 above)
Comment (elcallio) [reply src/net/tls.cc]: Removed

**Comment by elcallio** [comment]
ping?

**Comment by elcallio** [comment]
ping?

**Pull request merged**